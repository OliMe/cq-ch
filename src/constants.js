export const TYPE_COMMAND = 'command';

export const TYPE_QUERY = 'query';
