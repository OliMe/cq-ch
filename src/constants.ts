export const TYPE_COMMAND = 'command';

export const TYPE_QUERY = 'query';

export const IS_BROWSER = typeof window !== 'undefined';
