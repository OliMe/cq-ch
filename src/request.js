import getTransport from './event-transport/get-transport';
import { TYPE_QUERY } from './constants';
import { checkArguments } from './helpers/argument-checker';
import { channelCreatorCfg, requestChannelCfg } from './arguments.cfg';

/**
 * Declares a channel for sent queries in the service interface.
 * @param {Array} types Types of queries for sending.
 * @param {string} context Application context e.g. Namespace of command.
 * @return {Function} Function for sending queries to channel.
 */
export default function request (types, context) {
  checkArguments([types, context], 'request', channelCreatorCfg);
  return async (query, time = 200) => {
    checkArguments([query, time], 'requestChannel', requestChannelCfg);
    if (!types.includes(query.type)) {
      throw new TypeError('Trying to send query with type not in interface.');
    }
    query = { ...query, context };
    return await new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        reject('Time to answer exhausted.');
      }, time);
      getTransport(TYPE_QUERY).trigger(query.type, {
        ...query, resolve: (value: any) => {
          clearTimeout(timeout);
          return resolve(value);
        },
      });
    });
  };
}
