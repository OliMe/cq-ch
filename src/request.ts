import getTransport from './event-transport/get-transport';
import { TYPE_QUERY } from './constants';
import { Context, Message, OutputQuery, Types } from './types';
import { checkChannelCreator, checkRequestChannel } from './helpers/argument-checkers';

/**
 * Declares a channel for sent queries in the service interface.
 * @param {Array} types Types of queries for sending.
 * @param {string} context Application context e.g. Namespace of command.
 * @return {Function} Function for sending queries to channel.
 */
export default function request(types: Types, context: Context) {
  checkChannelCreator('request', types, context);
  return async function requestChannel<TResponse>(query: Message<TResponse>, time = 200) {
    checkRequestChannel('requestChannel', query, time);
    if (!types.includes(query.type)) {
      throw new TypeError('Trying to send query with type not in interface.');
    }
    return await new Promise((resolve: (value: TResponse) => void, reject) => {
      const timeout = setTimeout(() => {
        reject(
          `Time to answer exhausted. Message type: ${query.type}, request context: ${context}.`,
        );
      }, time);
      const extendedQuery: OutputQuery<TResponse> = {
        ...query,
        context,
        resolve: (value: TResponse) => {
          clearTimeout(timeout);
          resolve(value);
        },
      };
      getTransport(TYPE_QUERY).trigger(extendedQuery.type, extendedQuery);
    });
  };
}
