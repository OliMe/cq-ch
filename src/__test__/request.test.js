import request from '../request';
import getTransport from '../event-transport/get-transport';
import { TYPE_QUERY } from '../constants';

jest.mock('../event-transport/get-transport', () => {
  const trigger = jest.fn();
  return jest.fn(() => ({
    trigger,
  }));
});

describe('request', () => {
  it('creates channel successfully.', () => {
    expect(request(['test_type'], 'test')).toBeInstanceOf(Function);
  });
  it('creates channel unsuccessfully.', () => {
    const createChannelWithoutArguments = () => request();
    const createChannelWithFirstIncorrectArgument = () => request('not array');
    const createChannelWithFirstArgumentEmptyArray = () => request([], 'test');
    const createChannelWithoutSecondArgument = () => request(['test_type']);
    const createChannelWithSecondIncorrectArgument = () => request(['test_type'], 200);
    expect(createChannelWithoutArguments).toThrowErrorMatchingSnapshot();
    expect(createChannelWithFirstIncorrectArgument).toThrowErrorMatchingSnapshot();
    expect(createChannelWithFirstArgumentEmptyArray).toThrowErrorMatchingSnapshot();
    expect(createChannelWithoutSecondArgument).toThrowErrorMatchingSnapshot();
    expect(createChannelWithSecondIncorrectArgument).toThrowErrorMatchingSnapshot();
  });
});

describe('created by request function', () => {
  const channel = request(['first_type', 'second_type'], 'test');
  it('throws error when incorrect arguments passed', async () => {
    await expect(channel()).rejects.toThrowErrorMatchingSnapshot();
    await expect(channel({})).rejects.toThrowErrorMatchingSnapshot();
    await expect(channel({ type: 'third_type' })).rejects.toThrowErrorMatchingSnapshot();
  });
  it('sends request', () => {
    channel({ type: 'first_type' });
    expect(getTransport).toHaveBeenCalledWith(TYPE_QUERY);
    const { trigger } = getTransport();
    expect(trigger).toHaveBeenCalledWith('first_type', {
      type: 'first_type',
      context: 'test',
      resolve: expect.any(Function),
    });
  });
  it('receives respond', () => {
    expect.assertions(1);
    const resovedBy = 'test value';
    channel({ type: 'first_type' }).then(value => {
      expect(value).toBe(resovedBy);
    });
    const { trigger } = getTransport();
    trigger.mock.calls[0][1].resolve(resovedBy);
  });
  it('throws error on wait timeout expires', () => {
    jest.useFakeTimers();
    expect(channel({ type: 'first_type' }, 100)).rejects.toThrowErrorMatchingSnapshot();
    jest.runAllTimers();
  });
});
