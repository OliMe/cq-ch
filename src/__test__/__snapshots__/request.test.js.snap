// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`created by request function throws error on wait timeout expires 1`] = `undefined`;

exports[`created by request function throws error when incorrect arguments passed 1`] = `"Failed to execute 'requestChannel': first argument must be an Object with defined property type."`;

exports[`created by request function throws error when incorrect arguments passed 2`] = `"Failed to execute 'requestChannel': first argument must be an Object with defined property type."`;

exports[`created by request function throws error when incorrect arguments passed 3`] = `"Trying to send query with type not in interface."`;

exports[`request creates channel unsuccessfully. 1`] = `"Failed to execute 'request': first argument must be an Array with at least one element."`;

exports[`request creates channel unsuccessfully. 2`] = `"Failed to execute 'request': first argument must be an Array with at least one element."`;

exports[`request creates channel unsuccessfully. 3`] = `"Failed to execute 'request': first argument must be an Array with at least one element."`;

exports[`request creates channel unsuccessfully. 4`] = `"Failed to execute 'request': second argument must be a string."`;

exports[`request creates channel unsuccessfully. 5`] = `"Failed to execute 'request': second argument must be a string."`;
